framework/packages/abstract-core/src/native/ibc_host.rs

```diff
@@ -0,0 +1,164 @@
+//! # Abstract Api Base
+//!
+//! `abstract_core::adapter` implements shared functionality that's useful for creating new Abstract adapters.
+//!
+//! ## Description
+//! An Abstract adapter contract is a contract that is allowed to perform actions on a [proxy](crate::proxy) contract.
+//! It is not migratable and its functionality is shared between users, meaning that all users call the same contract address to perform operations on the Account.
+//! The api structure is well-suited for implementing standard interfaces to external services like dexes, lending platforms, etc.
+
+use crate::{
+    manager,
+    objects::{account::AccountId, chain_name::ChainName},
+};
+use cosmwasm_schema::QueryResponses;
+use cosmwasm_std::{Addr, CosmosMsg};
+
+pub mod state {
+    use cw_storage_plus::{Item, Map};
+
+    use crate::objects::ans_host::AnsHost;
+
+    use super::*;
+
+    /// Maps a chain name to the proxy it uses to interact on this local chain
+    pub const CHAIN_PROXYS: Map<&ChainName, Addr> = Map::new("ccl");
+    pub const REVERSE_CHAIN_PROXYS: Map<&Addr, ChainName> = Map::new("reverse-ccl");
+    /// Configuration of the IBC host
+    pub const CONFIG: Item<Config> = Item::new("cfg");
+
+    // Temporary structure to hold actions to be executed after account creation
+    pub const TEMP_ACTION_AFTER_CREATION: Item<ActionAfterCreationCache> = Item::new("act");
+
+    /// The BaseState contains the main addresses needed for sending and verifying messages
+    #[cosmwasm_schema::cw_serde]
+    pub struct Config {
+        /// AnsHost contract struct (address)
+        pub ans_host: AnsHost,
+        /// Address of the account factory, used to create remote accounts
+        pub account_factory: Addr,
+        /// Address of the local version control, for retrieving account information
+        pub version_control: Addr,
+    }
+
+    #[cosmwasm_schema::cw_serde]
+    pub struct ActionAfterCreationCache {
+        pub client_proxy_address: String,
+        pub account_id: AccountId,
+        pub action: HostAction,
+        pub chain_name: ChainName,
+    }
+}
+/// Used by Abstract to instantiate the contract
+/// The contract is then registered on the version control contract using [`crate::version_control::ExecuteMsg::ProposeModules`].
+#[cosmwasm_schema::cw_serde]
+pub struct InstantiateMsg {
+    /// Used to easily perform address translation on the app chain
+    pub ans_host_address: String,
+    /// Used to create remote abstract accounts
+    pub account_factory_address: String,
+    /// Version control address
+    pub version_control_address: String,
+}
+
+#[cosmwasm_schema::cw_serde]
+pub struct MigrateMsg {}
+
+#[cosmwasm_schema::cw_serde]
+#[non_exhaustive]
+pub enum InternalAction {
+    /// Registers a new account from a remote chain
+    Register {
```

Comment:
I think it would be really cool to add two things here. 

1: The ability to provide what modules to install on creation as well as to set a base account (all the fields that the create_account endpoint exposes on the factory)

2: The ability to trigger actions on any set of traces. I.e. I want to create an ICAA on Juno and Terra but there's no direct connection from Terra to my chain. So I call `Register`` with a set of traces on which I want to register accounts.

Another usecase is calling an action on a module that's installed on all my interchain accounts. 

The idea would be to have something like:

```rust
struct InterchainAbstractAccountAction {
    // HashSet of traces that the action should be performed for.
    // When the host receives this it constructs the trace for the current environment (trace of client + local chain name)
    // If that trace is in the set, execute the action (could be registration or any type of action)
    // If traces is not empty, forward the call to the local client which will forward it to the other chains if possible.
    traces: HashSet<AccountTrace>
    // Action to take on each trace for this account.
    action: HostAction
}
```

So basically we want to solve the problem of "I have an action I want to perform on chain X, Y and Z in one transaction."

1. Define the action (HostAction)
2. Define the chains (Traces)
3. Execute what you can locally and propagate the rest